hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -764,6 +764,56 @@ object LowerTableIR {
                 ++ rValueFields.map(f => f -> GetField(rEltRef, f)))
           })
 
+        case x@TableUnion(children) =>
+          val lowered = children.map(lower)
+          val keyType = x.typ.keyType
+          val newPartitioner = RVDPartitioner.generate(keyType, lowered.flatMap(_.partitioner.rangeBounds))
+          val repartitioned = lowered.map(_.repartitionNoShuffle(newPartitioner))
+
+          TableStage(
+            repartitioned.flatMap(_.letBindings),
+            repartitioned.flatMap(_.broadcastVals),
+            repartitioned.head.globals,
+            newPartitioner,
+            zipIR(repartitioned.map(_.contexts), ArrayZipBehavior.AssumeSameLength) { ctxRefs =>
+              MakeTuple.ordered(ctxRefs)
+            },
+            ctxRef =>
+              StreamMultiMerge(repartitioned.indices.map(i => repartitioned(i).partition(GetTupleElement(ctxRef, i))), keyType.fieldNames)
+          )
+
+        case x@TableMultiWayZipJoin(children, fieldName, globalName) =>
+          val lowered = children.map(lower)
+          val keyType = x.typ.keyType
+          val newPartitioner = RVDPartitioner.generate(keyType, lowered.flatMap(_.partitioner.rangeBounds))
+          val repartitioned = lowered.map(_.repartitionNoShuffle(newPartitioner))
+          val newGlobals = MakeStruct(FastSeq(
+            globalName -> MakeArray(lowered.map(_.globals), TArray(lowered.head.globalType))))
+          val globalsRef = Ref(genUID(), newGlobals.typ)
+
+          val keyRef = Ref(genUID(), keyType)
+          val valsRef = Ref(genUID(), TArray(children.head.typ.rowType))
+          val projectedVals = ToArray(mapIR(ToStream(valsRef)) { elt =>
```

Comment:
we may want to rewrite the StreamZipJoin semantics to avoid this in the future, if it's a performance bottleneck.

Reply:
Yeah, I forgot I had wanted to try making `curVals` be a stream, in which case I don't think this has any overhead vs. projecting out the keys in the StreamZipJoin emit.